### PR TITLE
Use variadic templates for rpc serialization

### PR DESCRIPTION
**Pull request by xf6wang**
Replaces variable argument lists in serialization with variadic templates. Since we no longer need to predefine a schema for rpc messages sent, we can change from hard coded callbacks to a registration system. RPC protocol has been modified such that a 32-bit header is always sent before any parameters. The leading 8 bits indicate the type (ie int, string ...) and the last 24 bits indicate the length in bytes. 

Needed by #3505.

- serialization functions are type-safe at compile time. 
- The main benefit of doing this is that the hard coded schema for `requests` and `responses` in `NetworkMessage.cc` is no longer needed. 
- requests/responses are now parsed in the same manner.

### CODE REVIEW

File: mgmt/api/NetworkMessage.h [hunk 1]
```diff
@@ -72,23 +72,111 @@ struct mgmt_message_sender {
...
+  if (!snd.is_connected()) {
+    return TS_ERR_NET_ESTABLISH; // no connection.
+  }
+
+  msglen = mgmt_message_length(params...);
+  msgbuf = (char *)ats_malloc(msglen + MGMT_HDR_LENGTH + 4);
+
+  // Add data header
+  memcpy((char *)msgbuf, &MGMT_DATA_HDR, MGMT_HDR_LENGTH);
+  memcpy((char *)msgbuf + MGMT_HDR_LENGTH, &msglen, 4);
```
Comment (xf6wang) [inline mgmt/api/NetworkMessage.h]: This isn't the cleanest. We're essentially building a `MgmtMarshallData` object from scratch. But, this entire function should go away in #3505.

File: mgmt/LocalManager.cc [hunk 2]
```diff
@@ -578,19 +578,18 @@ LocalManager::handleMgmtMsgFromProcesses(MgmtMessageHdr *mh)
...
+    if (mgmt_message_parse(data_raw, mh->data_len, &mparent, &mchild, &options) != -1) {
+      configFiles->configFileChild(mparent, mchild, (unsigned int)options);
     } else {
       mgmt_log("[LocalManager::handleMgmtMsgFromProcesses] "
                "MGMT_SIGNAL_CONFIG_FILE_CHILD mgmt_message_parse error\n");
     }
     // Output pointers are guaranteed to be NULL or valid.
-    ats_free_null(parent);
-    ats_free_null(child);
+    ats_free_null(mparent);
```
Comment (SolidWallOfCode) [inline mgmt/LocalManager.cc]: Outside the scope of this PR, but `MgmtMarshallString` should probably be updated to use `ats_scoped_str` or `std::unique_ptr` internally.

File: mgmt/ProcessManager.cc [hunk 3]
```diff
@@ -218,14 +218,14 @@ ProcessManager::reconfigure()
 void
 ProcessManager::signalConfigFileChild(const char *parent, const char *child, unsigned int options)
 {
-  static const MgmtMarshallType fields[] = {MGMT_MARSHALL_STRING, MGMT_MARSHALL_STRING, MGMT_MARSHALL_INT};
+  MgmtMarshallInt mgmtopt    = options;
+  MgmtMarshallString mparent = const_cast<MgmtMarshallString>(parent);
```
[the remark this answers is not in the document]
Comment (xf6wang) [reply mgmt/ProcessManager.cc]: It gives an assignment error from the compiler. I think `const MgmtMarshallString` gets interpreted as `char * const` since `MgmtMarshallString` is just an alias for `char *`.

File: mgmt/api/NetworkMessage.h [hunk 4]
```diff
@@ -72,23 +72,111 @@ struct mgmt_message_sender {
 };
 
 // Marshall and send a request, prefixing the message length as a MGMT_MARSHALL_INT.
-TSMgmtError send_mgmt_request(const mgmt_message_sender &snd, OpType optype, ...);
-TSMgmtError send_mgmt_request(int fd, OpType optype, ...);
+template <typename... Params>
+TSMgmtError
+send_mgmt_request(const mgmt_message_sender &snd, OpType optype, Params... params)
```
[the remark this answers is not in the document]
Comment (xf6wang) [reply mgmt/api/NetworkMessage.h]: Updated. The marshalling functions actually all use pointers. Is this still a big concern in terms of copying parameters?

File: mgmt/api/NetworkMessage.h [hunk 5]
```diff
@@ -72,23 +72,111 @@ struct mgmt_message_sender {
...
+{
+  ats_scoped_mem<char> msgbuf;
+  MgmtMarshallInt msglen;
+
+  if (!snd.is_connected()) {
+    return TS_ERR_NET_ESTABLISH; // no connection.
+  }
+
+  msglen = mgmt_message_length(params...);
```
[the remark this answers is not in the document]
Comment (xf6wang) [reply mgmt/api/NetworkMessage.h]: Updated. Used `std::forward`

File: mgmt/utils/MgmtMarshall.h [hunk 6]
```diff
@@ -61,17 +71,100 @@ struct MgmtMarshallData {
...
+MgmtMarshallInt mgmt_message_length(MgmtMarshallString *field);
+MgmtMarshallInt mgmt_message_length(MgmtMarshallData *field);
+MgmtMarshallInt mgmt_message_length(); // to allow for expansion
+
+template <typename T, typename... Rest>
+MgmtMarshallInt
+mgmt_message_length(T first, Rest... rest)
+{
+  MgmtMarshallInt len = mgmt_message_length(first);
+  return (len != -1) ? len + mgmt_message_length(rest...) : len;
```
[the remark this answers is not in the document]
Comment (xf6wang) [reply mgmt/utils/MgmtMarshall.h]: Updated to check both `first` and `rest` separately.

**Comment by SolidWallOfCode** [comment]
Overall I'm a bit concerned about using a single byte. That saves bandwidth but it's unclear to me that's the limiting resource here since these communications are presumed to be on the same host. On the downside this creates misalignments on most data reads which fortunately no longer cause segmentation faults on modern hardware but do nevertheless cost more than an aligned read. On the third hand it's just one read per integer into aligned storage so maybe it's premature optimization. I need to think about that.

File: mgmt/LocalManager.cc [hunk 7]
```diff
@@ -578,19 +578,18 @@ LocalManager::handleMgmtMsgFromProcesses(MgmtMessageHdr *mh)
...
-    char *parent                           = nullptr;
-    char *child                            = nullptr;
-    MgmtMarshallInt options                = 0;
-    if (mgmt_message_parse(data_raw, mh->data_len, fields, countof(fields), &parent, &child, &options) != -1) {
-      configFiles->configFileChild(parent, child, (unsigned int)options);
+    MgmtMarshallInt options = 0;
+    MgmtMarshallString mparent;
+    MgmtMarshallString mchild;
+    if (mgmt_message_parse(data_raw, mh->data_len, &mparent, &mchild, &options) != -1) {
+      configFiles->configFileChild(mparent, mchild, (unsigned int)options);
```
Comment (SolidWallOfCode) [inline mgmt/LocalManager.cc]: `static_cast<unsigned int>(options)`

File: mgmt/LocalManager.cc [hunk 8]
```diff
@@ -578,19 +578,18 @@ LocalManager::handleMgmtMsgFromProcesses(MgmtMessageHdr *mh)
...
   // Congestion Control - end
   case MGMT_SIGNAL_CONFIG_FILE_CHILD: {
-    static const MgmtMarshallType fields[] = {MGMT_MARSHALL_STRING, MGMT_MARSHALL_STRING, MGMT_MARSHALL_INT};
-    char *parent                           = nullptr;
-    char *child                            = nullptr;
-    MgmtMarshallInt options                = 0;
-    if (mgmt_message_parse(data_raw, mh->data_len, fields, countof(fields), &parent, &child, &options) != -1) {
-      configFiles->configFileChild(parent, child, (unsigned int)options);
+    MgmtMarshallInt options = 0;
+    MgmtMarshallString mparent;
```
Comment (SolidWallOfCode) [inline mgmt/LocalManager.cc]: Why is `options` initialized but not `MgmgMarshallString`? Does that type self initialize?

**Comment by SolidWallOfCode** [comment]
Thinking back, I thought I'd suggested using a 32-bit header for each item, with 8 bits of type and 24 bits of length (which should be sufficient for all the expected data). Did you do that and I missed it?

**Comment by xf6wang** [comment]
@SolidWallOfCode I didn't go with the 32-bit header for each item. My thinking was that for integer and long types, it wasn't necessary to send bits for length information since it would already be known.  I went with sending a single byte header to indicate the type. If this creates more problems with alignment I can change it to go with the 32-bit header.

**Comment by xf6wang** [comment]
Go big or go home right :) Updated to send a 32 bit header. I'm hesitant about having length indicated for integral types. I think in almost all situations `MgmtMarshallInt` is just used. This seems more explicit and it much easier than having to initialize each time with something like `MgmtMarshallIntegral val {some int, 4}`. It also makes assignments easier.

**Comment by zwoop** [comment]
[approve ci clang-format]

File: mgmt/api/NetworkMessage.h [hunk 9]
```diff
@@ -23,6 +23,7 @@
 
 #pragma once
 
+#include <tuple>
```
Comment (SolidWallOfCode) [inline mgmt/api/NetworkMessage.h]: Why is this included?

File: mgmt/utils/MgmtMarshall.h [hunk 10]
```diff
@@ -61,17 +75,112 @@ struct MgmtMarshallData {
...
+ssize_t mgmt_message_parse(const void *buf, size_t len); // to allow for expansion
+
+template <typename T, typename... Rest>
+ssize_t
+mgmt_message_parse(const void *buf, size_t len, T first, Rest &&... rest)
+{
+  if (buf == nullptr) {
+    return -1;
+  }
+  ssize_t nfirst = mgmt_message_parse(buf, len, first);
```
Comment (SolidWallOfCode) [inline mgmt/utils/MgmtMarshall.h]: Correct but for performance there's no point in doing the second parse if the first doesn't work. E.g. something more like
```
ssize_t n1, n2;
return -1 != (n1 = mgmt_message_parse(buf, len, first)) &&
    -1 != mgmt_message_parse(buff + n1, len -n1, std::forward<Rest>(rest)...)) ? n1 + n2 : -1;
```

File: mgmt/utils/MgmtMarshall.cc [hunk 11]
```diff
@@ -37,22 +39,67 @@ union MgmtMarshallAnyPtr {
...
+static constexpr size_t MGMT_LEN_MASK = 0x00FFFFFF;
+
+// For integral values, we ignore the length so we can create constants for the headers.
+static const uint32_t MGMT_INT_HDR  = (MGMT_INT_TYPE << MGMT_LEN_BITS);
+static const uint32_t MGMT_LONG_HDR = (MGMT_LONG_TYPE << MGMT_LEN_BITS);
+
+static uint32_t
+get_type_from_hdr(const MgmtMarshallHdr hdr)
+{
+  return (hdr >> MGMT_LEN_BITS); // shift away the length bits.
```
Comment (SolidWallOfCode) [inline mgmt/utils/MgmtMarshall.cc]: `return` is not a function.

File: mgmt/utils/MgmtMarshall.h [hunk 12]
```diff
@@ -61,17 +75,112 @@ struct MgmtMarshallData {
...
+ssize_t
+mgmt_message_read(int fd, T first, Rest &&... rest)
+{
+  ssize_t nfirst = mgmt_message_read(fd, first);
+  ssize_t nrest  = mgmt_message_read(fd, std::forward<Rest>(rest)...);
+  return (nfirst != -1 && nrest != -1) ? nfirst + nrest : -1;
+}
+/// mgmt_message_read ---------------------------------------------------------
+
+/// mgmt_message_write --------------------------------------------------------
```
Comment (SolidWallOfCode) [inline mgmt/utils/MgmtMarshall.h]: It's unclear to me why there's both `..._write` and `..._marshal`. If the message is marshalled to a buffer, what is the use of per type write methods?

File: mgmt/api/NetworkMessage.h (hunk 9 above)
Comment (xf6wang) [reply mgmt/api/NetworkMessage.h]: no reason. removed.

File: mgmt/utils/MgmtMarshall.h (hunk 10 above)
Comment (xf6wang) [reply mgmt/utils/MgmtMarshall.h]: updated.

File: mgmt/utils/MgmtMarshall.cc (hunk 11 above)
Comment (xf6wang) [reply mgmt/utils/MgmtMarshall.cc]: if you believe hard enough it is. removed parentheses

File: mgmt/utils/MgmtMarshall.h (hunk 12 above)
Comment (xf6wang) [reply mgmt/utils/MgmtMarshall.h]: So you have the option of either marshaling into a buffer or directly writing on the socket. It's more for flexibility. 

For example, `NetworkMessage.h` marshalls the whole message and sends it as a data object. The point of this is to attach a length so the entire marshalled message can be read into a buffer. This is necessary because the entire buffer is passed to the corresponding handler which figures out how to parse the buffer. Contrast this to a response which can directly send the parameters using `mgmt_message_write`.

**Comment by SolidWallOfCode** [comment]
It was determined at the Spring 2018 summit to take the different approach of using a third party serialization library.
